framework-plugins/lisk-framework-chain-connector-plugin/src/chain_connector_plugin.ts

```diff
@@ -311,8 +423,140 @@ export class ChainConnectorPlugin extends BasePlugin<ChainConnectorPluginConfig>
 		return false;
 	}
 
-	// eslint-disable-next-line @typescript-eslint/no-empty-function
-	private async _createCCU() {}
+	private async _calculateInboxUpdate(
+		sendingChainID: Buffer,
+		chainAccount: ChainAccount,
+	): Promise<InboxUpdate> {
+		// TODO: Fetch as many CCMs as configured in _ccuFrequency.ccm, between lastCertificateHeight and certificateHeight - After #7569
+		const serialzedCCMs = this._chainConnectorState.crossChainMessages
+			.slice(0, this._ccuFrequency.ccm)
+			.map(ccm => codec.encode(ccmSchema, ccm));
+
+		const outboxKey = this._rawStateStoreKey(sendingChainID);
+
+		const stateProveResponse = await this._sidechainAPIClient.invoke<ProveResponse>('state_prove', {
+			stateRoot: chainAccount.lastCertificate.stateRoot,
+			queries: [outboxKey],
+		});
+		const inclusionProofOutboxRoot: OutboxRootWitness = {
+			bitmap: stateProveResponse.proof.queries[0].bitmap,
+			siblingHashes: stateProveResponse.proof.siblingHashes,
+		};
+
+		const inboxUpdate: InboxUpdate = {
+			crossChainMessages: serialzedCCMs,
+			messageWitnessHashes: [],
+			outboxRootWitness: inclusionProofOutboxRoot,
+		};
+
+		return inboxUpdate;
+	}
+
+	private _rawStateStoreKey(chainID: Buffer) {
+		const moduleIDBuffer = Buffer.alloc(4);
+		moduleIDBuffer.writeInt32BE(MODULE_ID_INTEROPERABILITY, 0);
+		const storePrefixBuffer = Buffer.alloc(2);
+		storePrefixBuffer.writeUInt16BE(STORE_PREFIX_OUTBOX_ROOT, 0);
+
+		return Buffer.concat([chain.DB_KEY_STATE_STORE, moduleIDBuffer, storePrefixBuffer, chainID]);
+	}
+
+	private async _validateCertificate(
+		certificateBytes: Buffer,
+		certificate: Certificate,
+		blockHeader: chain.BlockHeader,
+		chainAccount: ChainAccount,
+	): Promise<CertificateValidationResult> {
+		const result: CertificateValidationResult = {
+			status: false,
+			chainStatus: chainAccount.status,
+			certificate,
+		};
+
+		if (chainAccount.status === CHAIN_TERMINATED) {
+			return result;
+		}
+
+		if (certificate.height <= chainAccount.lastCertificate.height) {
+			return result;
+		}
+
+		const certificateLivenessValidationResult = await this._verifyLiveness(
+			chainAccount.networkID,
+			certificate.timestamp,
+			blockHeader.timestamp,
+		);
+
+		result.livenessValidationResult = certificateLivenessValidationResult;
+
+		if (!certificateLivenessValidationResult.status) {
+			return result;
+		}
+
+		if (chainAccount.status === CHAIN_ACTIVE) {
+			result.status = true;
+
+			return result;
+		}
+
+		const validatorData = this._chainConnectorState.validatorsHashPreimage.find(
```

Comment:
does it need undefined check in case there is no preimage found?

Reply:
yes its better to have this check here